### PR TITLE
[FLINK-2093][gelly] Added difference Method

### PR DESCRIPTION
**Pull request by shghatge**
Tasks given on 5th June:
Add a difference function to  Graph.java
Modify the docs 'gelly-guide.md'
Add the test case for difference() method to GraphMutationsITCase.java
Added a new public method removeVertices(DataSet<Vertex<K, VV>> v)  to Graph.Java

### CODE REVIEW

**Comment by andralungu** [comment]
Hi @shghatge, 

Apart from the minor cosmetic suggestions I made, everything looks well. 

@vasia, could you double check this? After Shivani simplifies the difference method a bit, I would call this good to merge. 

Thanks!

File: docs/libs/gelly_guide.md [hunk 1]
```diff
@@ -240,6 +240,7 @@ Graph<Long, Double, Double> networkWithWeights = network.joinWithEdgesOnSource(v
     <img alt="Union Transformation" width="50%" src="fig/gelly-union.png"/>
 </p>
 
+* <strong>Difference</strong>: Gelly's `difference()` method performs a difference on the vertex and edge sets of the input graphs. The resultant graph is formed by removing the vertices and edges from the graph that are common with the second graph.
```
Comment (vasia) [inline docs/libs/gelly_guide.md]: we can rephrase this a bit.. there is one input graph and no "second" graph... I guess you copied from the union description above (which should also be changed).

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 2]
```diff
@@ -1234,6 +1234,18 @@ public void coGroup(Iterable<Edge<K, EV>> edge, Iterable<Edge<K, EV>> edgeToBeRe
...
 
 	/**
+	 * Performs Difference on the vertex and edge sets of the input graphs
+	 * removes common vertices and edges. If a source/target vertex is removed, its corresponding edge will also be removed
+	 * @param graph the graph to perform difference with
+	 * @return a new graph where the common vertices and edges have been removed
+	 */
+	public Graph<K,VV,EV> difference(Graph<K,VV,EV> graph) throws java.lang.Exception{
+		DataSet<Vertex<K,VV>> removeVerticesData = graph.getVertices();
+		final List<Vertex<K,VV>> removeVerticesList = removeVerticesData.collect();
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: I don't think we should use `collect()` here.. Keep in mind that (1) `collect()` will trigger program execution and (2) should not be used to collect large DataSets and input graph might have lots of vertices.

**Comment by vasia** [comment]
Hi @shghatge! Thank you for the PR!
I left you a couple of comments. Let us know if you have questions :)

File: docs/libs/gelly_guide.md [hunk 3]
```diff
@@ -240,6 +240,7 @@ Graph<Long, Double, Double> networkWithWeights = network.joinWithEdgesOnSource(v
     <img alt="Union Transformation" width="50%" src="fig/gelly-union.png"/>
 </p>
 
+* <strong>Difference</strong>: Gelly's `difference()` method performs a difference on the vertex and edge sets of the input graphs. The resultant graph is formed by removing the common vertices and edges from the graph.
```
Comment (andralungu) [inline docs/libs/gelly_guide.md]: I think @vasia also wanted you to update the description for union ^^

Now, this still looks a bit unclear. It seems that there are two input graphs. You should make it obvious that the current graph gets differentiated with an input graph. That way, you won't leave room for comments.

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 4]
```diff
@@ -1151,6 +1151,23 @@ public void coGroup(Iterable<Vertex<K, VV>> vertex, Iterable<Vertex<K, VV>> vert
 		}
 	}
 
+
+	public Graph<K, VV, EV> removeVertices(DataSet<Vertex<K, VV>> verticesToBeRemoved){
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: Yes, this is definitely the idea, but right now you are duplicating a lot of code. Can we find a smarter way (i.e. that requires as little code duplication as possible) :)?

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 5]
```diff
@@ -1151,6 +1151,23 @@ public void coGroup(Iterable<Vertex<K, VV>> vertex, Iterable<Vertex<K, VV>> vert
 		}
 	}
 
+
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: always add Javadoc to new methods.

**Comment by andralungu** [comment]
Looks good to me. +1

**Comment by vasia** [comment]
Hi @andralungu, @shghatge,
I'm a bit confused by your changes.. Why did you add a new `removeVertices` API method for DataSet?
I thought we had agreed that we'll have add/remove* methods for Lists and union/difference for DataSets, no?

**Comment by andralungu** [comment]
Hi @vasia ,

In essence the `difference` method is just a fancy way of removing vertices, right?
When you remove a vertex, you also remove the edge for which it was a source/target.

Since the add/remove vertices methods work just for lists and collect is "unsafe", we mutually agreed to overload `removeVertices` to work for data sets. 
This way you would duplicate the least amount of code. Otherwise, you would take the exact code in the DataSet removeVertices and duplicate it in difference. That's not very practical IMO.

Also, it may occur that a user has a DataSet of elements to remove. An extra removeVertices won't really hurt then, would it? 

But if you have suggestions on how to improve this, we are more than eager to hear about them :)

-Andra

**Comment by vasia** [comment]
Hey,

difference in not a fancy way of removing vertices, no :)
If you recall our previous conversation in #678, we decided to have add/remove methods for small mutations (i.e. can be stored in a List)  and union/difference for large mutations (for large data in the form of Graph).

The `collect()` method is not "unsafe". It is simply not advised to be used for large datasets. If you think about it, it needs to retrieve the data (which might be distributed over several nodes) from to local JVM and trigger the program execution.

If you found that the only way to implement `difference` is to overload `removeVertices`, then we should also consider what this means in terms of the API. If we have a remove* method for DataSets, don't we also need a corresponding add*? You see how this goes back to #678 :)

**Comment by shghatge** [comment]
Hello @vasia 
Could you please elaborate on what should be the future course of action?
Should I put the content of removeVertices(DataSet<>) method in the Difference() method so as to not add a new public method? Or is it better to overload the add methods for DataSets  too?

**Comment by vasia** [comment]
Hi @shghatge!

For this issue, I think we should only add the `difference` method. If you want to avoid duplicating code, you can always define a private helper method.
If we decide that we need overloaded public methods for additions and removals, we should discuss this in a separate JIRA in my view.
Let me know if you have questions!

**Comment by shghatge** [comment]
Updated the PR by changing the removeVertices(DataSet<Vertex<K, VV>>) method  access from Public to Private. It is only used as a helper function for the difference method.

**Comment by andralungu** [comment]
Hi @shghatge ,

Don't forget to remove the definition for the public removeVertices(DataSet) from the documentation.  

Up for discussion: should we keep the name removeVertices for the private, helper method or should we call it something else, like removeVerticesAndEdges... Names are not my strongest point, but I guess you got the idea :) Personally, I am fine with the current name!

**Comment by shghatge** [comment]
Updated the docs accordingly.

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 6]
```diff
@@ -1234,6 +1248,17 @@ public void coGroup(Iterable<Edge<K, EV>> edge, Iterable<Edge<K, EV>> edgeToBeRe
 	}
 
 	/**
+	 * Performs Difference on the vertex and edge sets of the input graphs
+	 * removes common vertices and edges. If a source/target vertex is removed, its corresponding edge will also be removed
+	 * @param graph the graph to perform difference with
+	 * @return a new graph where the common vertices and edges have been removed
+	 */
+	public Graph<K,VV,EV> difference(Graph<K,VV,EV> graph) throws java.lang.Exception{
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: why the `throws` declaration?

**Comment by vasia** [comment]
Hi @shghatge!
Thank you for the quick update. Apart from my minor comment, this looks good now :)

**Comment by shghatge** [comment]
Updated PR

**Comment by vasia** [comment]
Thank you @shghatge! I'll merge this :)

**Pull request closed without merging**